Vladi Toshev 82019

vladi-tech opened the pull request:
Added homework RCS1 and RCS 2;
Used Git CMD for upload;
Modified operator >> before pull request;

Commit:
Added Homeworks RCS1 and RCS2

Commit:
Modified >> operator

Commit:
RCS3 LList Reverse

vladi-tech commented:
Added RCS3 in the pull request.

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+{
+    strcpy(this->string, other.string);
+}
+String::String(const char *other_string) : string{new char[strlen(other_string)]}, _size{strlen(other_string)}, _capacity{strlen(other_string) * 2}
+{
+    strcpy(this->string, other_string);
+}
```
заделената памет и капацитета ми се разминават

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+{
+    return _size == 0;
+}
+const char String::at(const int &pos) const
+{
+    if (0 >= pos || pos > _size)
+    {
+        std::cout << std::endl
+                  << "Invalid position";
+        return '\0';
+    }
+
+    return string[pos];
+}
```
Няма много смисъл от това да извеждаш съобщение. Или assert-вай, или хвърли exception.

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+{
+    return string[this->_size - 1];
+}
+void String::append(const String &newString)
+{
+    while (this->_size + newString._size >= _capacity)
+    {
+        resize_String();
+    }
+
+    this->_size += newString._size;
+    strcat(this->string, newString.string);
+}
```
Сигурен ли си, че ще се преорамрези достатъчно?

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+    this->_size += newString._size;
+    strcat(this->string, newString.string);
+}
+char *String::c_str() const
+{
+    char *result = new char[this->_size + 1];
+    strcpy(result, string);
+    return result;
+}
```
const char* също става

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+    delete[] string;
+    string = newString;
+}
+void String::resize(const size_t &n)
+{
+    if (n < _size)
+    {
+        std::cout << std::endl
+                  << "Invalid new capacity size!";
+        return;
+    }
+    String new_String(this->string, n);
+    swapString(new_String);
+    new_String.~String();
+}
```
Моля те, недей да извикваш деструктора. Той сам ще се извика...

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+    return *this;
+}
+
+String &String::operator+=(const String &other)
+{
+    while (this->_size + other._size >= _capacity)
+    {
+        resize_String();
+    }
+    char *result = new char[this->_size + other._size];
+    result = strcat(this->string, other.string);
+    this->_size += other._size;
+    return *this;
+}
```
орязваш низа.
append би помогнал

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+    return string[pos];
+}
+
+String::operator bool() const
+{
+    return !(_size == 0);
+}
```
???
_size != 0 ?

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+}
+
+//Input, output operators
+std::ostream &operator<<(std::ostream &out, const String &object)
+{
+    if (object.empty() == 1)
+    {
+        out << "   ";
+    }
+    else
+    {
+        out << "String: " << object.string;
+    }
+    out << "  Size: " << object._size;
+    out << "  Capacity: " << object._capacity;
+    return out;
+}
```
Аз бих си спестил спомагателните съобщения

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.cpp:
```diff
@@ -0,0 +1,235 @@
+    out << "  Capacity: " << object._capacity;
+    return out;
+}
+std::istream &operator>>(std::istream &in, String &object)
+{
+    in >> object._capacity;
+    object.string = new char[object._capacity];
+    in >> object.string;
+    object._size = strlen(object.string);
+    object.string[object._size] = '\0';
+    return in;
+}
```
Добре е да се чете на buckets т.е. на части, докато има какво да се чете и да се добавят към низа.

NickGeorgiev commented on student_work/VladiToshev/RCS1/String.h:
```diff
@@ -0,0 +1,59 @@
+    String &operator+=(const String &);
+    char &operator[](const int &);
+    const char &operator[](const int &) const;
+    operator bool() const;
```
искаме само експлицитно конвертиране

NickGeorgiev commented on student_work/VladiToshev/RCS2/test.cpp:
```diff
@@ -0,0 +1,199 @@
+
+TEST_CASE("String")
+{
+    SUBCASE("Construction")
+    {
+        SUBCASE("Default construction")
+        {
+            String test_String;
+            CHECK(test_String.size() == 0);
+            CHECK(test_String.capacity() == 1);
+        }
+        SUBCASE("Construction with capacity")
+        {
+            String test_String(20);
+            CHECK(test_String.size() == 0);
+            CHECK(test_String.capacity() == 20);
+        }
+        SUBCASE("Construction with a given string")
+        {
+            String test_String("Test");
+            CHECK(test_String.size() == 4);
+            CHECK(test_String.capacity() == 8);
+        }
+        SUBCASE("Construction with given string and capacity")
+        {
+            String test_String("TestStr", 20);
+            CHECK(test_String.size() == 7);
+            CHECK(test_String.capacity() == 20);
+        }
+        SUBCASE("Copy constructor")
+        {
+            String test_String("Test", 15);
+            String copy(test_String);
+
+            CHECK(copy.size() == 4);
+            CHECK(copy.capacity() == 15);
+        }
+    }
```
Никъде не виждам да проверяваш съдържанието

NickGeorgiev commented on student_work/VladiToshev/RCS2/test.cpp:
```diff
@@ -0,0 +1,199 @@
+        }
+    }
+
+    SUBCASE("Destructor")
+    {
+        String test_String("Test", 15);
+        test_String.~String();
+
+        CHECK(test_String.size() == 0);
+        CHECK(test_String.capacity() == 0);
+    }
```
Извикването на деструктора...

NickGeorgiev commented on student_work/VladiToshev/RCS3/LList.cpp:
```diff
@@ -0,0 +1,145 @@
+    return value;
+}
+
+template <class T>
+void LList<T>::reverse()
+{
+    if (start == nullptr || start == end)
+    {
+        return;
+    }
+
+    LList<T> tempList;
+    tempList = *this;
+    this->destroyList();
+    while (tempList.start != nullptr && tempList.end != nullptr)
+    {
+        this->toStart(tempList.popStart());
+    }
+}
```
Идеята беше да се реализира без да се създава нов низ, а да се обърнат указателите. Така става бавно, а и може списъците да са твърде големи и да не се поберат заедно в паметта.

Commit:
Addressed changes to RCS1 and RCS2, modified RCS3, added additional tests

Commit:
Added requested changes to RCS1 and RCS2, modified RCS3

vladi-tech commented:
Modified the Linked List reverse function;

Addressed the requested changes in the RCS1 file;

Added additional tests in the String and LList unit test file

Commit:
Updated the unit tests visial layout with 'Given, When and Then' paragraphs, removed a few unnecessary tabs in LList.cpp
